### PR TITLE
Remove angular from business details

### LINKED ISSUE #9482: [Bye Bye Angular] Enterprise Settings -> 'Business Details'

### PR DESCRIPTION
**Pull request by binarygit**
#### What? Why?

Closes #9482  

#### What should we test?

- Login as enterprise manager/super admin
- In page `enterprises/edit` under `business details` tab
- Ascertain `what's this?` tooltips show correct text and hide themselves when you hover out
- Ascertain that you can add new Terms and Conditions file
- Ascertain that you are presented with a modal box and have to confirm before adding new T&C file
- Ascertain that you view the file after you add it and update
- Ascertain that you can delete the T&C file

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (binarygit) [inline app/webpacker/controllers/terms_and_conditions_controller.js:29]: I'll be first to admit that I think the code is a bit hacky. I couldn't find another way to add this functionality.

The modal window needs to open after the `browse` button is clicked by the user and only after the `confirm` button in the modal is clicked should the default behaviour (the file selection window open) continue.

In angular this behaviour is achieved through here:
```coffee
"app/assets/javascripts/admin/enterprises/directives/terms_and_conditions_warning.js.coffee"

     # When the user presses continue in the warning modal, we open the browser file input dialog
    scope.continue = ->
      scope.template.dialog('close')
      $rootScope.$evalAsync()
```

And I'm reading the whole thing as `event.preventDefault()` then if the user presses `continue` -> `event.continueDefault()` but there's no such thing in native js, which is why I came to this solution. 

Also I have a half a second's delay before clicking the input because when I didn't have that the file selection window wouldn't show. Must be something in the close-modal code that prevents it.

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (binarygit) [inline app/views/admin/enterprises/form/_business_details.html.haml:41]: I know you guys aren't very familiar with SR, so I thought I would link the relavent section in the docs to make it easier to understand what's happening.
https://docs.stimulusreflex.com/rtfm/morph-modes#scoping-page-morphs

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (jibees) [inline app/webpacker/controllers/terms_and_conditions_controller.js:28]: Why not using a `target`? Isn't the `#enterprise_terms_and_conditions` element in the terms_and_conditions_controller?

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (jibees) [reply app/webpacker/controllers/terms_and_conditions_controller.js:29]: I should re-read the code, and test myself. Will give a try.

**Review by jibees — COMMENTED** [review]
Thanks for this one. 
I need to pull the PR and deep dive into, to be sure to understand. Let's make that when I have time.
Also I wanted to have @mkllnk opinion on it.

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (binarygit) [reply app/webpacker/controllers/terms_and_conditions_controller.js:28]: I can't use `target` because the `#enterprise_terms_and_conditions` is in another instance of the t&c controller. This method is used by

```haml
%input.button.red{ type: 'button', value: t('js.admin.modals.continue'), "data-controller": "terms-and-conditions", "data-action": "click->help-modal#close click->terms-and-conditions#add" }
```
and this is also inside the t&c controller but in a different instance of it. It's controller comes from

```
# app/components/help_modal/help_modal_component.html.haml

%div{ id: @id, "data-controller": "help-modal", "data-action": "keyup@document->help-modal#closeIfEscapeKey" }
  .reveal-modal-bg.fade{ "data-help-modal-target": "background", "data-action": "click->help-modal#close" }
  
```

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (binarygit) [reply app/webpacker/controllers/terms_and_conditions_controller.js:29]: Sure. Take your time JB.

btw, below where you have commented on the segment of code

```ruby
  add() {
    let input = document.querySelector("#enterprise_terms_and_conditions");
    input.removeAttribute("data-action");
```

This line `input.removeAttribute("data-action");` is just removing the event handlers from the `input`. I am not sure that it 100% removes the event handlers but after the `data-action` attributes are removed the corresponding `stimulus` actions don't run which is why you can click on the input and have the file selection window pop up. In contrast to, having the `terms and conditions warning` modal popup.

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (jibees) [inline app/views/admin/enterprises/form/_business_details.html.haml:34]: Now there's only one `terms-and-conditions` controller which also encapsulate the modal content.

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (jibees) [inline app/views/admin/enterprises/form/_business_details.html.haml:47]: Here is the trick: we add a button and hide the file input.

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (jibees) [inline app/views/admin/enterprises/form/_business_details.html.haml:54]: The modal is now inside its controller: then the controller is called once on the page

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (jibees) [inline app/webpacker/controllers/terms_and_conditions_controller.js:12]: Each time the fileinput change, we update this little span that contains the file name: though we can simulate the fileinput behavior (ie. with the to-be-uploaded filename)

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (jibees) [inline app/webpacker/controllers/terms_and_conditions_controller.js:32]: And then, we just have to click ;)

**Review by jibees — CHANGES_REQUESTED** [review]
I had some comments, and then I made a proposal!
Let me know what you think about ;)

**Comment by binarygit** [comment]
wow thanks for taking the time to go through this and come up with this fantastic solution. I'll fix the failing tests and remove the extra code.

File: app/controllers/api/v0/terms_and_conditions_controller.rb
Comment (jibees) [inline app/controllers/api/v0/terms_and_conditions_controller.rb:16]: This one should be here since the API should still exists

File: config/routes/api.rb
Comment (jibees) [inline config/routes/api.rb:42]: Same, API still exists.

**Review by jibees — CHANGES_REQUESTED** [review]

File: app/reflexes/enterprise_edit_reflex.rb
Comment (binarygit) [reply app/reflexes/enterprise_edit_reflex.rb:4]: I'll have a look at this

**Comment by binarygit** [comment]
Hey @jibees could you take a look? I verify ability before letting users purge the T&C

**Review by jibees — APPROVED** [review]
Looks good to me, thanks a lot @binarygit !
💪

**Comment by binarygit** [comment]
Thanks for the great feedback JB :smiley: :trophy:

File: app/reflexes/enterprise_edit_reflex.rb
Comment (mkllnk) [inline app/reflexes/enterprise_edit_reflex.rb:10]: I think that it would be better to use `purge_later`.

File: app/reflexes/example_reflex.rb
Comment (mkllnk) [inline app/reflexes/example_reflex.rb:3]: Why do we need this example file? If you want the documentation then we could add that to the ApplicationReflex, maybe?

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (mkllnk) [reply app/webpacker/controllers/terms_and_conditions_controller.js]: > I'll be first to admit that I think the code is a bit hacky. I couldn't find another way to add this functionality.

I'm wondering if we can simplify the T&Cs upload model, @openfoodfoundation/train-drivers-product-owners. Can we show that information as normal text within the form so that you can read it any time instead of just popping it up when you try to upload a file?

I'm sure that once you read that warning once, it's annoying to have it getting in the way when you want to upload the next file. It's very unexpected to have a modal pop up like that. A bit of that is good to make sure that people don't overlook it but I think that it's a bit much. I'm hoping that we can improve UX and simplify code at the same time.

**Review by mkllnk — COMMENTED** [review]
Good work, team.

File: app/reflexes/example_reflex.rb
Comment (binarygit) [reply app/reflexes/example_reflex.rb:3]: I usually have this around so that one can just copy this file when creating a new reflex file. Just like how we have an `example_controller.js` in `app/webpacker/controllers`

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (lin-d-hop) [reply app/webpacker/controllers/terms_and_conditions_controller.js]: Conversation moved to Slack [here](https://openfoodnetwork.slack.com/archives/C018S101R4J/p1665983975954109).

**Review by mkllnk — APPROVED** [review]
I'm not sure how to proceed with this. I think that it would be better to remove the modal part and simplify the code. Lynne agreed on Slack that it would be better to have the information inline on the page instead of a modal. But you worked out a solution already in this could just go to testing and be merged.

So, to make things simple, I'll approve and move to test-ready but if testers find issues then it may be easier to remove complexity instead of trying to solve the issues with more complexity. Sounds good?

**Comment by mkllnk** [comment]
Actually, specs seem to fail consistently on CI. Redis errors?

**Comment by binarygit** [comment]
I agree, let's merge this for now. Should I squash the commits into a single one? `replace angular in business details tab`

**Comment by mkllnk** [comment]
We need to fix the specs first.

**Comment by binarygit** [comment]
```
Failure/Error: raise CannotConnectError, "Error connecting to Redis on #{location} (#{error.class})"
     
     Redis::CannotConnectError:
       Error connecting to Redis on localhost:6379 (Errno::ECONNREFUSED)
```

The specs pass locally but fail here. I do not know how to fix this. I searched but nothing relevant poped up.

**Comment by mkllnk** [comment]
* https://github.com/openfoodfoundation/openfoodnetwork/issues/9825

**Comment by binarygit** [comment]
Just pushed after squashing the commits. Thanks guys for the reviews.

**Comment by mkllnk** [comment]
> Just pushed after squashing the commits.

Okay, cool. Just as a general comment: you got a lot of changes in one commit now. Smaller commits with incremental changes are usually better and we don't tend to squash them in the end. Sometimes we remove detours and squash some fixup commits.

**Comment by drummer83** [comment]
@mkllnk Is the feedback-needed label still needed here?

**Comment by mkllnk** [comment]
No, this is ready for testing.

**Comment by filipefurtad0** [comment]
Hey @binarygit ,

Folowed your detailed notes on what to test and after logging in as an enterprise manager/super admin and visiting business details tab, checken whether:
- `what's this?` tooltips show correct text and hide themselves when you hover out - these need clicking, but this was the previous behavior already (current master). All good here.
- Ascertain that you can add new Terms and Conditions file - OK
- Ascertain that you are presented with a modal box and have to confirm before adding new T&C file - OK. Uploading a new file it triggers the same warning as when compared with master. A minor detail is the color of the header -> it is now displayed in red:

image

- Ascertain that you view the file after you add it and update - OK
- Ascertain that you can delete the T&C file - this is not working correctly, clicking remove triggers a Bugsnag error and leaves the file:
Notification [seen here](https://openfoodnetwork.slack.com/archives/G010VN35QU9/p1668173009182929):
image

image


Perhaps best to address this last issue before merging. Please let me know if you need access to Bugsnag or the pics above is sufficient to reproduce/fix the error locally.

Thanks for your work on this :muscle:

**Comment by binarygit** [comment]
Thanks for testing this @filipefurtad0 I will take a look and figure out what's wrong.

**Comment by binarygit** [comment]
@jibees I looked into this but can't figure out what is going wrong. 
I found this issue discussing this same problem: https://github.com/stimulusreflex/stimulus_reflex/issues/187 but unlike the OP inside our `cable.yml` our redis URL **is** set to the correct url.

```yml
 production:
  adapter: redis
  url: <%= ENV.fetch("OFN_REDIS_URL", "redis://localhost:6380/0") %>
  channel_prefix: your_application_production

staging:
  adapter: redis
  url: <%= ENV.fetch("OFN_REDIS_URL", "redis://localhost:6380/0") %>
  channel_prefix: your_application_staging
```

I'm not really sure what is hindering the WS to be established, so would really appreciate your help in this :smile:

**Comment by jibees** [comment]
@binarygit 
I'm wondering which error do you have exaclty... Is it the same than the one here: https://github.com/openfoodfoundation/openfoodnetwork/pull/9804 ?

**Comment by binarygit** [comment]
@jibees yup it is the same error

**Comment by jibees** [comment]
I suspect a nginx error, that does not _forward_ (not sure this is the exact term) the `ws` protocol due to a misconfiguration on staging (and production I guess). Maybe could @dacook have a idea?

**Comment by jibees** [comment]
Hi 👋 @binarygit 
This one has conflict with `app/reflexes/example_reflex.rb`. As I wanted to deploy it on `fr-staging` and small test it, I propose to rebase this branch.

**Comment by jibees** [comment]
I've tested this one, and it wasn't working because of:

```
actioncable.esm.js:231 WebSocket connection to 'wss://staging.coopcircuits.fr/cable' failed: 
```

I guess we missed 87dff339b66ec351f35aca68206e4bb1dd4386be
I'll cherry pick this one, and give another try.

**Comment by jibees** [comment]
I works!! 🎉

**Comment by RachL** [comment]
@jibees hellooo so we move it to code review or testing? I assumed you are done with staging FR so I'm going to use it a bit.

**Comment by jibees** [comment]
@RachL 
Actually, this one is ready and can be tested in staging fr but we need to review, test and provision https://github.com/openfoodfoundation/ofn-install/pull/852 first...

**Comment by jibees** [comment]
I've rebased, solved conflicts and made a minor change around ability. I think this one is ready for a review.

File: spec/system/admin/enterprises/terms_and_conditions_spec.rb
Comment (mkllnk) [inline spec/system/admin/enterprises/terms_and_conditions_spec.rb:34]: Cool, I didn't know `make_visible`. That's handy. But it also means that we are not testing the new functionality, right?

**Review by mkllnk — APPROVED** [review]
Wow, this is pretty cool. I'm still perplexed how this works. So the edit action renders the whole page. Then we stimulate a reflex but the reflex doesn't define anything to render. Does it render the whole edit action again and cuts the reflex root out? Whatever it does, it seems to work.

File: spec/system/admin/enterprises/terms_and_conditions_spec.rb
Comment (jibees) [reply spec/system/admin/enterprises/terms_and_conditions_spec.rb:34]: Yes, we don't test the full functionality. Actually, it's impossible because once the system file browser opens there is no way for the driver to interact with it.

**Comment by jibees** [comment]
> Then we stimulate a reflex but the reflex doesn't define anything to render. Does it render the whole edit action again and cuts the reflex root out?

As far as I understood: 
> By default, StimulusReflex updates your entire page.
[Source](https://docs.stimulusreflex.com/guide/morph-modes.html)

File: app/reflexes/enterprise_edit_reflex.rb
Comment (dacook) [inline app/reflexes/enterprise_edit_reflex.rb:7]: It seems unnecessary to break over two lines.

```suggestion
    throw :forbidden unless can?(:remove_terms_and_conditions, @enterprise)
    
```
I also suggest including parentheses for the `can?` method call.  I _love_ that paretheses are optional in Ruby, but  find that it's clearer to read when the code is always formatted the same way (especially when there's more than one parameter like here).

And, I've added blank line because this functions like other "guard clauses" (`raise`, `return`, etc)

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (dacook) [reply app/views/admin/enterprises/form/_business_details.html.haml:47]: Just wondering, is hiding the file input part of the new design? Is it a general rule that we wish to apply on other file inputs?

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (dacook) [inline app/views/admin/enterprises/form/_business_details.html.haml:51]: It seems to me that the "delete" button should be closer to the name of the file that it deletes.

Looking at [this screenshot](https://user-images.githubusercontent.com/49817236/201351670-ace8c9ed-87d8-47be-b952-d677de0bd33c.png) I found, I think it should be on the same line as the "upload" button, so it doesn't take up another line.

I think it's as simple as removing line 49 (`.pad-top`) and indenting lines 50-51. GitHub won't let me use the suggestion feature here.. oh well we can leave it as it is.

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (dacook) [inline app/webpacker/controllers/terms_and_conditions_controller.js:29]: I can't find any references to this method, is it a built-in name?

**Review by dacook — CHANGES_REQUESTED** [review]
Looking good, yay!

I just have a couple of suggestions, and a question.

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (jibees) [reply app/views/admin/enterprises/form/_business_details.html.haml:51]: I have another proposal: 

Capture d’écran 2023-03-08 à 10 18 34

I've made the remove action a link (not a button) and moved it closer to the filename (because it actually remove this one, not the just uploaded one)

What do you think?

File: app/webpacker/controllers/terms_and_conditions_controller.js
Comment (jibees) [reply app/webpacker/controllers/terms_and_conditions_controller.js]: Good catch, removed!

File: app/views/admin/enterprises/form/_business_details.html.haml
Comment (dacook) [reply app/views/admin/enterprises/form/_business_details.html.haml]: So much better!

**Review by dacook — APPROVED** [review]
Great improvement!

**Comment by drummer83** [comment]
Hi @binarygit and @jibees,

I had a look at this PR. Deleting T&C files works fine now. However there are some regressions around design and confirmation.

| Test case | Test result |
| - | - |
| Tooltips | :heavy_check_mark: |
| Add new T&C file | :heavy_check_mark: |
| View uploaded T&C file | :heavy_check_mark: |
| Delete T&C file | :question: Button is gone... Confirmation is gone... (see video below) |

Left: **BEFORE PR**; Right: **AFTER PR**

https://user-images.githubusercontent.com/1790176/226206529-752c6e66-93b9-4703-a153-f167c7892a58.mp4

## Result
Once again adding feedback-needed label. We do have regressions on design (no button for 'remove file', just text) and confirmations (no confirmation after deleting the T&C file.
Waiting for feedback from @openfoodfoundation/train-drivers-product-owners.

**Comment by RachL** [comment]
thanks for the detailed testing @drummer83 !

I don't see this as a big problem for now. Even without the confirmation text, the user can see the file is not there anymore. And I never understood why we needed a big red button to remove the file here... What's more confusing to me it's that's probably one of the few actions we don't need to save :joy:

I would merge as is and improve the design later with upcomping styleguide, it's not a feature people use every day. But happy if someone else thinks otherwise!

**Pull request merged**
